reduce ES hits in paginator from 2 to 1

PetrDlouhy opened the pull request:
The paginator initiates COUNT requests to ElasticSearch, that are not necessary, because the count is returned by the full query.
This modification takes the result count from the query and uses it for validation resulting in only on ES hit per request.

barseghyanartur commented on src/django_elasticsearch_dsl_drf/pagination.py:
```diff
@@ -53,12 +53,11 @@ def page(self, number):
-        number = self.validate_number(number)
         bottom = (number - 1) * self.per_page
         top = bottom + self.per_page
-        if top + self.orphans >= self.count:
```
In any case, this shouldn't have been removed.

```python
if top + self.orphans >= self.count:
    top = self.count
```

PetrDlouhy replied:
I think it should be removed or rewritten. The `self.count` raises the unwanted ES hit.
Regarding my testing, it works just fine - the ES is inteligent enough not to cause problems when the `top` is higher than element count.

But maybe I have missed something - i.e. I don't know, what is `self.orphans`.

PetrDlouhy replied:
OK, I see the failed tests in Travis, so I will take look at them.

PetrDlouhy replied:
I am still not sure, what is the problem with the failing tests - I expect, that the problem is, that it gets different ES data structure than on my local environment.

I was not able to run the tests locally, but I will try further tomorrow.

barseghyanartur replied:
Regarding removal of orphans.

[The minimum number of items allowed on the last page, defaults to zero. Use this when you don’t want to have a last page with very few items. If the last page would normally have a number of items less than or equal to orphans, then those items will be added to the previous page (which becomes the last page) instead of leaving the items on a page by themselves. For example, with 23 items, per_page=10, and orphans=3, there will be two pages; the first page with 10 items and the second (and last) page with 13 items.](https://github.com/django/django/blob/master/django/core/paginator.py#L79)

In general, although it's a technically good optimisation fix for Elasticsearch, it does break some commonly accepted functionality. Additionally, page number validation is messed up with.

I fixed error in your branch [here](https://github.com/barseghyanartur/django-elasticsearch-dsl-drf/commit/5f92add7d31fee35140559def62b8fb7a581bd6e).

I would accept additional paginator classes though, but not as defaults. Let them be `QueryFriendlyPaginator`, `QueryFriendlyPageNumberPagination` and `QueryFriendlyLimitOffsetPagination`. Probably `validate_number` shall be overwritten for these classes too (to avoid the count query).

I'll leave this up to you.

PetrDlouhy replied:
@barseghyanartur Thank you for the fix.

Regarding the orphans - I would fix this behaviour by requesting the orphaned nodes in additional request. So the number of ES hits will be 1 in most cases, but with if there are orphans, it would fall back to 2 requests. That way it would be beneficial for most users.
Do you think it is a good solution?

barseghyanartur reviewed (changes requested):
See my comments [here](https://github.com/barseghyanartur/django-elasticsearch-dsl-drf/pull/225/files#r559828765)

PetrDlouhy commented:
I have improved the code and added tests, that ensure the behaviour. Here are some notes:

- I have fixed the failing test in @barseghyanartur code - it was caused by conflicting name of `get_count` method.
- I have solved the orphanes problem by adding one more request to ES only in the case of last page with orphanes.
   - Although, I am not sure, if the result as list would not cause problems elsewhere. I am not sure, if we are able to merge two `Response` objects, but if not, we could re-request all the items by `object_list = self.object_list[self.count].execute()`, but it could have lower performance in the rare cases of orphans usage.
- I realised, that the URL query parameter `page_size` in `test_pagination.py` does nothing, so I have added default value to the `PageNumberPagination` to match the behaviour with examples in it's apidocs.
- I had to add there also `orphans_query_param` parameter in order to be able to test orphans.
- I have added testing number of ES queries to `test_pagination.py`.

barseghyanartur commented:
@PetrDlouhy:

Thanks. I'll test it asap.

barseghyanartur commented:
@PetrDlouhy:

Sorry for the delay. I merged it into the [225-reduce-ES-hits-in-paginator-from-2-to-1](https://github.com/barseghyanartur/django-elasticsearch-dsl-drf/tree/225-reduce-ES-hits-in-paginator-from-2-to-1) branch and made some changes.

For now, I don't want to make the optimised pagination standard. Instead, I made it optional. See the following [example](https://github.com/barseghyanartur/django-elasticsearch-dsl-drf/blob/225-reduce-ES-hits-in-paginator-from-2-to-1/examples/simple/search_indexes/viewsets/book/query_friendly_pagination.py):

```python
from django_elasticsearch_dsl_drf.pagination import QueryFriendlyPageNumberPagination
from .default import BookDocumentViewSet


class QueryFriendlyPaginationBookDocumentViewSet(BookDocumentViewSet):

    pagination_class = QueryFriendlyPageNumberPagination
```

Same with tests. I separated the two (fill fix the overlaps later).

This will be released as soon as all the tests pass along with some other pending fixes/deprecations.

Thanks so much for your time and efforts!
